blaze/expr/tests/test_expr.py

```diff
@@ -187,3 +188,69 @@ def test_pickle_roundtrip():
     t = symbol('t', 'var * int64')
     expr = (t + 1).mean()  # some expression with more than one node.
     assert expr.isidentical(pickle.loads(pickle.dumps(expr)))
+
+
+def test_coalesce():
+    # check case where lhs is not optional
+    s = symbol('s', 'int32')
+    t = symbol('t', 'int32')
+    expr = coalesce(s, t)
+    assert expr.isidentical(s)
+
+    # check case where lhs is null dshape
+    u = symbol('u', 'null')
+    expr = coalesce(u, s)
+    assert expr.isidentical(s)
+
+    # check optional lhs non-optional rhs
+    v = symbol('v', '?int32')
+    expr = coalesce(v, s)
+    # rhs is not optional so the expression cannot be null
+    assert_dshape_equal(expr.dshape, dshape('int32'))
+    assert expr.lhs.isidentical(v)
+    assert expr.rhs.isidentical(s)
+
+    # check optional lhs non-optional rhs with promotion
+    w = symbol('w', 'int64')
+    expr = coalesce(v, w)
+    # rhs is not optional so the expression cannot be null
+    # there are no either types in datashape so we are a type large enough
+    # to hold either result
+    assert_dshape_equal(expr.dshape, dshape('int64'))
+    assert expr.lhs.isidentical(v)
+    assert expr.rhs.isidentical(w)
+
+    # check optional lhs and rhs
+    x = symbol('x', '?int32')
+    expr = coalesce(v, x)
+    # rhs and lhs are optional so this might be null
+    assert_dshape_equal(expr.dshape, dshape('?int32'))
+    assert expr.lhs.isidentical(v)
+    assert expr.rhs.isidentical(x)
+
+    # check optional lhs and rhs with promotion
+    y = symbol('y', '?int64')
+    expr = coalesce(v, y)
+    # rhs and lhs are optional so this might be null
+    # there are no either types in datashape so we are a type large enough
+    # to hold either result
+    assert_dshape_equal(expr.dshape, dshape('?int64'))
+    assert expr.lhs.isidentical(v)
+    assert expr.rhs.isidentical(y)
```

Comment:
Some error / edge cases for testing:
* non-scalar dshapes: `Record`, others?;
* non-integer dshapes: `string` comes to mind;
* non-atomic expression arguments with compatible dshapes.